### ISSUE TITLE
AMS should error when no serializer can be found

### ISSUE THREAD

**Issue by cantino** [rails-api/active_model_serializers#1853]
I think the default behavior of AMS should be to error if a serializer cannot be found.  Otherwise, it's easy to stumble into a major security hole.

Imagine you have a controller, maybe something like:

```ruby
class RecentUsersController < ApplicationController
  def index
    render json: User.recent.limit(5)
  end
end
```

and a serializer:

```ruby
class UserSerializer < ActiveModel::Serializer
  attribute :first_name
  attribute :user_since
end
```

Now, imagine you either put that serializer at the wrong path by accident, or you named it `RecentUserSerializer`, but forgot to specify it by name in the render call.  Either way, you're now calling `as_json` on the models and handing hashed passwords out over your API.

I'm using something like this to fix the issue, but I think it should be default behavior:

```ruby
module ErrorOnMissingSerializer
  module AsJSONOverride
    def as_json(*_)
      serializable_resource = ActiveModelSerializers::SerializableResource.new(self)
      if serializable_resource.serializer
        serializable_resource.as_json
      else
        raise NotImplementedError, "No serializer was found for #{self.class.name}"
      end
    end
  end

  class Railtie < Rails::Railtie
    ActiveSupport.on_load :active_record do
      ActiveRecord::Base.include ErrorOnMissingSerializer::AsJSONOverride
    end
  end
end
```

**Comment by bf4** [comment]
@cantino I think there is room for a 'strict mode', that raises an `MissingSerializer = Class.new(ArgumentError)`  instead of returning `nil` when no serializer is found.  The complication is that we'd need to ensure that the object respond to `read_attribute_for_serialization` or something so that it's not a string, hash, array, etc.

Relevant code https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model/serializer.rb#L81-L87

```diff
    # @api private
    # Find a serializer from a class and caches the lookup.
    # Preferentially returns:
    #   1. class name appended with "Serializer"
    #   2. try again with superclass, if present
-    #   3. nil
+    #   3. config.models_require_serializer? ? fail(MissingSerializer) : nil
    def self.get_serializer_for(klass)
      return nil unless config.serializer_lookup_enabled
      serializers_cache.fetch_or_store(klass) do
        # NOTE(beauby): When we drop 1.9.3 support we can lazify the map for perfs.
        serializer_class = serializer_lookup_chain_for(klass).map(&:safe_constantize).find { |x| x && x < ActiveModel::Serializer }

        if serializer_class
          serializer_class
        elsif klass.superclass
          get_serializer_for(klass.superclass)
+        elsif klass.respond_to?(:read_attribute_for_serialization) && config.models_require_serializer?
+           fail MissingSerializer
+        else
+          nil
        end
      end
    end
```

**Comment by bf4** [comment]
FWIW, if you use a a JSON-Schema to validate your requests/responses in tests, you can be assured you're not rendering a model without a serializer...

**Comment by NullVoxPopuli** [comment]
@GregPK if you want to turn that in to a PR, we can get some more formal discussions on it? :-)

**Comment by GregPK** [comment]
@NullVoxPopuli True - let's discuss with a concrete example :point_up:

**Comment by bf4** [comment]
somewhat related: 
- https://github.com/rails-api/active_model_serializers/issues/1500 
- https://github.com/rails-api/active_model_serializers/issues/1813
- https://github.com/rails-api/active_model_serializers/pull/1767
- https://github.com/rails-api/active_model_serializers/issues/1442
- https://github.com/rails-api/active_model_serializers/pull/1757
- https://github.com/rails-api/active_model_serializers/pull/1883
- https://github.com/rails-api/active_model_serializers/issues/1743
- https://github.com/rails-api/active_model_serializers/issues/1337
- https://github.com/rails-api/active_model_serializers/pull/1436
- https://github.com/rails-api/active_model_serializers/pull/1636

**Comment by rromanchuk** [comment]
I just got burned by this, still trying to figure out why `serializer_for` is being called four times, twice with an empty model, but this caused my `User` object to call `as_json` leaking internal data, without any indication. 

https://gist.github.com/rromanchuk/36701818f977c6b1b0325db519048754

**Comment by mhluska** [comment]
The `as_json` hack above doesn't seem to work in the latest version (0.10.7). How can I enforce specifying a serializer for a model today?

**Comment by bf4** [comment]
@mhluska 

> How can I enforce specifying a serializer for a model today?

That's actually a different feature from OP's "AMS should error when no serializer can be found"

I don't remember when    `self.get_serializer_for(klass)` is called, but you could play with prepending it with a module 

```diff
def self.get_serializer_for(klass)
+   fail ArgumentError, "Serializer lookup is disabled" unless config.serializer_lookup_enabled
  super
```

feel free to open a new issue for that.

**Comment by sergioisidoro** [comment]
+1

**Comment by yungtrizzle** [comment]
Yeah this monkey patch is broken for me on version 0.10.14 + rails 6.1
